### ISSUE TITLE
Fetching metadata from yt-dlp using --add-metadata

### ISSUE THREAD

**Issue by pncnmnp** [deepjyoti30/ytmdl#272]
# Feature Request

## Checklist

- [X] I've searched the bugtracker for similar feature requests including closed ones.
- [X] I will add the enhancement label to the issue.

## Description

Hey, I have been using this library in the development of [Phoenix10.1](https://github.com/pncnmnp/phoenix10.1) for a while now, and I have noticed that for many cases there is a mismatch between the songs fetched from Youtube and their corresponding metadata for that song.

So, today morning I decided to understand why this is happening, and how we could improve upon the existing state of `ytmdl`.

When I download anything locally using `yt-dlp`, I use a CLI command similar to:
```bash
yt-dlp --extract-audio --add-metadata --embed-thumbnail --audio-format mp3 -o "%(title)s.%(ext)s" --postprocessor-args "-id3v2_version 3"
```

This ensures that the metadata attached to that song is also fetched and stored in the id3v2 format. It's nothing fancy, but it does ensure that the album name, artist name, and song name are correct for most cases.

Now, when I created a `yt-dlp.conf` file:
```
# Enable the add-metadata option
--add-metadata
```

and passed the path as a parameter to the `args.ytdl_config`, I noticed that in many cases, the original metadata was accurate, and the resulting `ytmdl` logic would overwrite this data.

At this point, I tried using `--skip-meta`, as I was under the assumption, that this would not tamper the original metadata from `yt-dlp`. 

But that was not the case. It seems like `ytmdl` will not encode the fetched metadata from `yt-dlp` in the final mp3 file when `skip_meta` is `True`.

I also tried approaching this problem from a different perspective, by seeing if [there is a way to get the Youtube `URL` that `ytmdl` used for fetching a particular song](https://github.com/deepjyoti30/ytmdl/blob/a9251c8e047eca64649b70821c81bb33c5895183/ytmdl/main.py#LL272C5-L272C9). However, it seems like this is not supported, and the only way to achieve this would be to manually parse the log files.

In conclusion, I am trying to understand how I could keep the metadata fetched from `yt-dlp` using the `--add-metadata` tag without it ever being overwritten.

Based on my investigation, this is not currently supported, and I feel that it would be nice to have this feature included in `ytmdl`. It won't help us get a vast amount of metadata information, but having a correct album name, artist name, and song name is a good start. We can then use this to accurately fetch data from itunes, musicbrainz, etc.

Cross-referencing a relevant Github Issue from Phoenix10.1: https://github.com/pncnmnp/phoenix10.1/issues/17

**Comment by pncnmnp** [comment]
My apologies, I should have closely examined the code of `ytmdl`. Fortunately, there is a workaround for the above-mentioned issue: directly call the `ytmdl.core.search` method.
```python
import ytmdl
args = ytmdl.main.arguments()
args.SONG_NAME = ["SONG_NAME"]
args.choice = 1
args.quiet = True

url, title = ytmdl.core.search(args.SONG_NAME[0], args)
```

The URL can then be utilized directly with `yt-dlp`. I will experiment with this and provide an update soon. Thank you.

**Comment by deepjyoti30** [comment]
The above is correct. You can use the core functions directly and establish various functionalities. Moreover, I think the feature you requested is still a good one since the metadata that yt-dlp provides can be a fallback meta. It doesn't look good but it will be something.

The above `yt-dlp` code looks good. You can refer to how `ytmdl` is using `yt-dlp` by checking the `ytmdl/yt.py` file.

BTW, [Phoenix10.1](https://github.com/pncnmnp/phoenix10.1) looks like a good idea, will try it out!

**Comment by pncnmnp** [comment]
@deepjyoti30 I just experimented with this code and it seems to work surprisingly well for our use cases, especially when compared to `ytmdl`’s metadata search.

```python
import ytmdl
import yt_dlp
import eyed3
import glob
import os
import nltk

args = ytmdl.main.arguments()
args.SONG_NAME = ["YOUR_SONG_NAME"]
args.choice = 1
args.quiet = True

url, title = ytmdl.core.search(args.SONG_NAME[0], args)

ydl_opts = {
    'format': 'bestaudio/best',
    'postprocessors': [{
        'key': 'FFmpegExtractAudio',
        'preferredcodec': 'mp3',
        'preferredquality': '192',
    },
    {'key': 'FFmpegMetadata'},],
    'outtmpl': '%(title)s.%(ext)s',
}

with yt_dlp.YoutubeDL(ydl_opts) as ydl:
    error_code = ydl.download([url])

song_file = glob.glob("*.mp3")[0]
metadata = eyed3.load(song_file)
fetched_artist = metadata.tag.artist

os.remove(song_file)

infos = ytmdl.metadata.get_from_itunes(args.SONG_NAME[0])
most_accurate = sorted([info.json for info in infos[:10]], key=lambda x: nltk.edit_distance(x["artistName"], fetched_artist))[0]

print(most_accurate)
```

I will try to conduct more thorough testing on thousands of songs to verify this claim.

**Comment by pncnmnp** [comment]
@deepjyoti30, I feel confident in providing my comment on the above-mentioned code now.

TLDR; **The code mentioned above appears to fetch metadata with about 29% more accuracy than the default logic of `ytmdl`**.

I began by searching for datasets containing song titles and artist names. I found an excellent one on Kaggle: [Top Spotify songs from 2010-2019 - BY YEAR](https://www.kaggle.com/datasets/leonardopena/top-spotify-songs-from-20102019-by-year).

Thus, I made minor enhancements to the aforementioned script and compared it to the metadata created by `ytmdl.main.main`.

Using the suggested script resulted in 82.8% accuracy, whereas `ytmdl` had about 54%. This finding aligns with our observations on Phoenix10.1, where 1/4 to 1/3 of all songs display incorrect metadata. 

The dataset comprises roughly 600 songs, although I couldn't test the entire set as it was quite slow. Nonetheless, after analyzing 250 songs, I obtained the above results that should provide a suitable estimate.

Based on this experiment, I am confident that the best action for Phoenix10.1 is to utilize `--add-metadata` with `yt-dlp` to collect the song's actual metadata. Following this, we will extract further information from `itunespy` using an `edit_distance` based method. 

We will continue to use `ytmdl.core.search` as it has proven quite practical.

[Here is the code and data for the entire experiment](https://gist.github.com/pncnmnp/730b5229ae40d02e123938e37d8bf90a).

**Comment by pncnmnp** [comment]
@deepjyoti30, we have successfully made the changes mentioned above to Phoenix 10.1. 

If you plan on implementing these changes to ytmdl, check out https://github.com/pncnmnp/phoenix10.1/pull/18 for inspiration. 
Thank you for your help, and please feel free to ask any questions about our implementation.

**Comment by hariprasanths** [comment]
Failling back to yt-dlp's metadata when metadata fetch errors out would help a lot. I would prefer to have an option to not skip a track altogether when metadata fetch fails.

**Comment by deepjyoti30** [comment]
@hariprasanths Support for `youtube` as a fallback for metadata is added in the above commit. I will include it in the next release.
